torch/distributed/_sharding_spec/api.py

```diff
@@ -84,6 +105,40 @@ def _verify_dim(dim):
         if not (isinstance(dim, int) or isinstance(dim, str)):
             raise ValueError(f'{dim} needs to either be an int or str')
 
+    def shard(self, tensor_sizes: List[int], process_group=None) -> List[ShardMetadata]:
```

Comment:
This API design forces us to have a process group and have process groups enabled?
The issue is that those constraints make it not usable for planning or testing purposes.

This API should allow us to answer questions like "how different is rank 0 usage if I go from 14 to 12 workers?"

Reply:
I see, yeah good point. I updated the PR with two new APIs now:
1. ShardingSpec.build_metadata, this will return a ShardedTensorMetadata, I will remove the process_group argument as it seems not very useful. This can be used by planning or testing purposes. 
2. ShardingSpec.shard, this will return an actual ShardedTensor, which will have the process_group as argument as we really scattering the shards to different ranks.

For answering the question like "how different is rank 0 usage if I go from 14 to 12 workers?" I feel like this might be the responsibility of a "ShardingPlanner", ShardingSpec should just try to build the metadata base on the spec (list of placements) given. ShardingPlanner can explore different options to make use of the resources (i.e. device usage). We will build a Customizable ShardingPlanner as described in this RFC https://github.com/pytorch/pytorch/issues/72138, as user can define their own rules to explore their cluster set up.